### PR TITLE
Add snacraft config

### PR DESCRIPTION
**Pull request by musoke**
This allows the easy generation of snaps as requested in #227.

This commit does not add automatic generation and publication of snaps
to the Snap Store, but they can easily be generated from source for 
local/personal use by

```
snapcraft
sudo snap install gutenberg_xxx.snap
```

I have not taken the liberty of registering or publishing the app in the Snap Store.  If that is done, updates can be pushed to the store automatically from travis.

### CODE REVIEW

**Comment by Keats** [comment]
Thanks!

Should this live outside the repo though? I don't use Ubuntu so I'm unlikely to maintain it myself, I only maintain the AUR repo and I have a separate repo for that (it's required by AUR anyway)

**Comment by Keats** [comment]
What do you think @musoke ?

**Comment by musoke** [comment]
Sorry for the delay @Keats.  Personally I would keep it in the repo so anyone who wants a snap can easily build it themselves.  I don't foresee the need for any maintenance other than changing the version number.  Even that could be eliminated by setting it to something like `git`.

**Comment by musoke** [comment]
Also: Canonical intends that snaps become ubiquitous on distros other than Ubuntu.  Personally I use them mostly to runs apps not yet on Debian stable.

**Comment by Keats** [comment]
I don't mind letting in the repo if I don't really need to do anything, but how would snap knows how to build Gutenberg from the file in the PR or even find this snap file?
Basically what are the steps needed from this PR to a snapcraft user being able to install Gutenberg from snap?

**Comment by musoke** [comment]
I believe that the simplest way to publish a snap to Ubuntu's store would be to [use their build service](https://build.snapcraft.io/).  It seemed easy when I tried it just now, but didn't finish as I didn't think it was appropriate for me to register the package name.  Builds would by default be marked as unstable, so you would not personally be making a stability guarantee to users whom install the snap.

Alternatively, one can [use Travis](https://docs.snapcraft.io/build-snaps/ci-integration). That may be less desirable 

- it looks like gutenberg's build times are already 30 minutes
- you would have to do a bit more than register the snap name
- may involve some maintenance

There is also Launchpad, which I use for a project of my own.  However, this would probably be more work for you.

**Comment by Keats** [comment]
Thanks, I'll try their build service to see how that looks

**Comment by Keats** [comment]
I found https://github.com/theoriginalbigbaddave/Cobalt/blob/master/snap/snapcraft.yaml which is very similar except for the `source` and `source-tag` info and the build-packages: are those not needed?

Sadly installing snapcraft didn't work for my distro so I'll have to test it directly via their build service (I registered the name), any chance you can confirm it builds locally with this snapcraft.yml?

File: snap/snapcraft.yaml [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+  gutenberg:
+    command: gutenberg
+    plugs:
+      - network
+      - network-bind
+
+parts:
+  gutenberg:
+    plugin: rust
+    rust-channel: nightly
```
Comment (Keats) [inline snap/snapcraft.yaml]: Gutenberg builds on stable, nightly is not 100% safe

**Review by Keats — CHANGES_REQUESTED** [review]

**Comment by musoke** [comment]
I took a look at the snapcraft docs. It seems like `source` is only need when pulling in remote packages, so I don't understand why cobalt uses it.

I just pushed an update to use the stable rust channel.

I tested building this - works for me on Debian Stretch with snapcraft 2.42.1.

**Comment by Keats** [comment]
Cool I'll merge it and try to build it with their service later today, thanks!

**Comment by Keats** [comment]
It fails on the build server: https://build.snapcraft.io/user/Keats/gutenberg/239694

**Comment by musoke** [comment]
I followed the instructions on that page. It starts compiling but runs into a problem with a custom build command for `sass-sys`.  Here is truncated output.

```
   Compiling term-painter v0.2.4
   Compiling rand v0.3.22
   Compiling bytes v0.4.6
   Compiling quote v0.5.1
   Compiling thread_local v0.3.5
   Compiling debug_unreachable v0.1.1
   Compiling syn v0.11.11
   Compiling log v0.3.9
   Compiling idna v0.1.4
   Compiling chrono v0.4.2
   Compiling mio v0.6.14
error: failed to run custom build command for `sass-sys v0.4.4`
process didn't exit successfully: `/root/build_gutenberg/parts/gutenberg/build/target/release/build/sass-sys-d9b94e18146f841e/build-script-build` (exit code: 101)
--- stderr
fatal: Not a git repository (or any of the parent directories): .git
thread 'main' panicked at 'error running make: Os { code: 2, kind: NotFound, message: "No such file or directory" }', libcore/result.rs:945:5
note: Run with `RUST_BACKTRACE=1` for a backtrace.

warning: build failed, waiting for other jobs to finish...
error: failed to compile `gutenberg v0.3.3 (file:///root/build_gutenberg/parts/gutenberg/build)`, intermediate artifacts can be found at `/root/build_gutenberg/parts/gutenberg/build/target`

Caused by:
  build failed
Failed to run '/root/build_gutenberg/parts/gutenberg/rust/bin/cargo install -j4 --root /root/build_gutenberg/parts/gutenberg/install --path /root/build_gutenberg/parts/gutenberg/build' for 'gutenberg': Exited with code 101.
Verify that the part is using the correct parameters and try again.
Debug mode enabled, dropping into a shell
root@snapcraft-idolisingly-unbigamous-erich:~#
```

I guess we need to tell snapcraft to build libsass first?  _Should_ be relatively straightforward, but we will see.

**Comment by musoke** [comment]
Actually, maybe snapcraft just didn't clone the libsass submodule.
Either way, I find it odd that `snapcraft cleanbuild --debug` fails although `snapcraft` works.

**Comment by Keats** [comment]
Is there a forum somewhere where we can get support? It's a bit hard for me to debug without having snapcraft installed

**Comment by musoke** [comment]
There is a forum: https://forum.snapcraft.io/categories

I just checked again, the pre-merge branch musoke/snap also fails to build with `snapcraft cleanbuild --debug`.  `snapcraft` is fine.

**Comment by Keats** [comment]
Ok, I'll post something on the forum later today. Thanks!

**Comment by Keats** [comment]
Having had a second look, it looks like `snapcraft cleanbuild` does not pull the submodules maybe? Which ends up with having an empty directory in libsass and git failing as it's missing the submodule

**Comment by musoke** [comment]
Yes, I think it is exactly that.

**Comment by Keats** [comment]
Is there no way to build the snap using `snapcraft` rather than cleanbuild?

**Comment by musoke** [comment]
When I was testing locally I just ran `snapcraft` and it worked no 
probelem. However, it seems that snapcraft's build service uses 
`snapcraft cleanbuild`.

**Comment by Keats** [comment]
Posted on https://forum.snapcraft.io/t/snap-of-gutenberg/6080

**Pull request merged**